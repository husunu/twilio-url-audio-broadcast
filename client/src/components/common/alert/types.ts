export enum AlertType {
  SUCCESS = 'success',
  INFO = 'info',
  ERROR = 'error',
  WARNING = 'warning',
}

export type AlertDto = {
  message?: string;
  type?: AlertType;
};
