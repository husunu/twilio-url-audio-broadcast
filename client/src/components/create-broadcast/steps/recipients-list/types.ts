import { INVALID_NUMBERS, VALID_NUMBERS } from './constants';

export type ValidatedNumber = {
  line: number;
  number: string;
  issue?: string | null;
};

export type ValidatedNumbers = {
  [INVALID_NUMBERS]: ValidatedNumber[];
  [VALID_NUMBERS]: ValidatedNumber[];
};
