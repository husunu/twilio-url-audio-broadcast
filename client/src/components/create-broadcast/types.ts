import React from 'react';
import { PropTypes } from '@material-ui/core';

import { StepType, ConfigureStep, RecipientsStep, ReviewStep } from '@/types';
import { ValidatedNumber } from '@/components/create-broadcast/steps/recipients-list/types';

export type StepItem = {
  key: StepType;
  label: string;
  icon: React.ReactNode;
  location: string;
  component: React.FunctionComponent<StepProps>;
  next?: string;
  prev?: string;
  nextButtonColor?: PropTypes.Color;
  nextButtonText?: string;
  nextButtonDisclaimer?: string;
};

export type StepProps = {
  data: ConfigureStep | RecipientsStep | ReviewStep;
  step: StepItem;
};

type NumbersTable = {
  headers: string[];
  rows: Array<keyof ValidatedNumber>;
};

export const VALID_NUMBERS_TABLE: NumbersTable = {
  headers: ['Line', 'Recipients Phone Number'],
  rows: ['line', 'number'],
};

export const INVALID_NUMBERS_TABLE: NumbersTable = {
  headers: ['Line', 'Number', 'Issue'],
  rows: ['line', 'number', 'issue'],
};
