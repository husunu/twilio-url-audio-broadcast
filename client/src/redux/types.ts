import {
  COMPLETE_CONFIGURE_STEP,
  COMPLETE_RECIPIENTS_STEP,
  COMPLETE_REVIEW_STEP,
  DISCARD_BROADCAST,
} from './constants';

import { ConfigureStep, RecipientsStep, ReviewStep } from '@/types';

export type CompleteConfigureStep = {
  type: typeof COMPLETE_CONFIGURE_STEP;
  payload: ConfigureStep;
};

export type CompleteRecipientsStep = {
  type: typeof COMPLETE_RECIPIENTS_STEP;
  payload: RecipientsStep;
};

export type CompleteReviewStep = {
  type: typeof COMPLETE_REVIEW_STEP;
  payload: ReviewStep;
};

export type Steps = {
  [key: string]: ConfigureStep | RecipientsStep | ReviewStep;
};

export type CreateBroadcastState = {
  steps: Steps;
};

export type DiscardBroadcast = {
  type: typeof DISCARD_BROADCAST;
  payload?: any;
};

export type CreateBroadcastActionType =
  | CompleteConfigureStep
  | CompleteRecipientsStep
  | CompleteReviewStep
  | DiscardBroadcast;
