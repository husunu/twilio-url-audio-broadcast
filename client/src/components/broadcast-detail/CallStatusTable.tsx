import React, { useCallback } from 'react';
import { PAGINATOR_TEST_ID, RECIPIENTS_TABLE_TEST_ID } from './constants';
import { BroadcastMeta, Call } from '@/types';
import { Setter } from '@/types/setter';
import { useRowsPerPageOptions } from '@/hooks/useRowsPerPageOptions';

import {
  LinearProgress,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TablePagination,
  TableRow,
} from '@material-ui/core';

type CallStatusTableProps = {
  recipients: Call[];
  meta: BroadcastMeta;
  loading: boolean;
  pageCount: number;
  rowsPerPage: number;
  setRowsPerPage: Setter<number>;
  currentPage: number;
  setCurrentPage: Setter<number>;
};

export const CallStatusTable: React.FC<CallStatusTableProps> = ({
  recipients,
  meta,
  loading,
  pageCount,
  rowsPerPage,
  setRowsPerPage,
  currentPage,
  setCurrentPage,
}) => {
  const rowsPerPageOptions = useRowsPerPageOptions(meta?.total);

  const onChangeRowsPerPage = useCallback(
    (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      const rows = parseInt(event.target.value, 10);
      setRowsPerPage(rows);
      setCurrentPage(0);
    },
    [setRowsPerPage, setCurrentPage]
  );

  const onChangePagination = useCallback(
    (event: unknown, value: number) => {
      setCurrentPage(value);
    },
    [setCurrentPage]
  );

  return (
    <>
      {recipients && recipients.length > 0 && (
        <>
          <TableContainer
            component={Paper}
            style={{
              marginBottom: loading ? '0' : '4px',
            }}
          >
            <Table
              aria-label="Notification recipients calls details and statuses"
              size="small"
              data-testid={RECIPIENTS_TABLE_TEST_ID}
            >
              <TableHead>
                <TableRow>
                  <TableCell>Number</TableCell>
                  <TableCell>Status</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {recipients?.map((recipient) => (
                  <TableRow key={recipient.callSid}>
                    <TableCell>{recipient.to}</TableCell>
                    <TableCell>{recipient.status}</TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
          {loading && <LinearProgress />}
        </>
      )}
      {pageCount > 0 && meta.total && (
        <TablePagination
          data-testid={PAGINATOR_TEST_ID}
          rowsPerPageOptions={rowsPerPageOptions}
          component="div"
          page={currentPage}
          count={meta.total}
          rowsPerPage={rowsPerPage}
          onChangePage={onChangePagination}
          onChangeRowsPerPage={onChangeRowsPerPage}
        />
      )}
    </>
  );
};
