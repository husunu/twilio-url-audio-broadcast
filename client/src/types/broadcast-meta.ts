import { CallStatus } from './call-status';

export type Count = {
  queued: number;
  initiated: number;
  ringing: number;
  'in-progress': number;
  completed: number;
  busy: number;
  'no-answer': number;
  canceled: number;
  failed: number;
};

export type BroadcastMeta = {
  total?: number;
  completed?: number;
  count?: Count;
  broadcastStatus?: CallStatus;
};
