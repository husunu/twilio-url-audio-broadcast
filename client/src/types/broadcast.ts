import { Channel } from './channel';

export type Broadcast = {
  broadcastId: string;
  friendlyName: string;
  message: string;
  from: string;
  channel: Channel;
  dateCreated: Date;
  canceled: boolean;
  completed: boolean;
  endDate?: Date;
};
