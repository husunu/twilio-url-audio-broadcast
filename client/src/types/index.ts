export * from './call';
export * from './broadcast';
export * from './broadcast-meta';
export * from './call-status';
export * from './channel';
export * from './step-type';
export * from './steps';
export * from './setter';
