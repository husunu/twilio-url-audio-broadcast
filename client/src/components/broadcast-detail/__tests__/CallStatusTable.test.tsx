import React from 'react';
import { render } from '@testing-library/react';
import { CallStatusTable } from '../CallStatusTable';
import { Call, CallStatus } from '@/types';

const defaultProps = {
  recipients: [] as Call[],
  meta: { total: 0 },
  loading: false,
  pageCount: 0,
  rowsPerPage: 10,
  setRowsPerPage: jest.fn(),
  currentPage: 1,
  setCurrentPage: jest.fn(),
};

describe('BroadcastDetail > CallStatusTable', () => {
  test('renders correctly', () => {
    const props = {
      ...defaultProps,
      recipients: [
        {
          callSid: 'CS000000',
          to: '+100001',
          status: CallStatus.IN_PROGRESS,
        },
      ],
      meta: { total: 1 },
      pageCount: 1,
      currentPage: 0,
    };

    const { asFragment } = render(<CallStatusTable {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders nothing when there are no recipients', () => {
    const { asFragment } = render(<CallStatusTable {...defaultProps} />);
    expect(asFragment()).toMatchInlineSnapshot(`<DocumentFragment />`);
  });
});
