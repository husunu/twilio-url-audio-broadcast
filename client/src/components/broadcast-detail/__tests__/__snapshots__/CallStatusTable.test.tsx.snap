// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BroadcastDetail > CallStatusTable renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
    style="margin-bottom: 4px;"
  >
    <table
      aria-label="Notification recipients calls details and statuses"
      class="MuiTable-root"
      data-testid="recipients-table"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
            scope="col"
          >
            Number
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
            scope="col"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
          >
            +100001
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
          >
            in-progress
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="MuiTablePagination-root"
    data-testid="paginator"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        class="MuiTablePagination-spacer"
      />
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        1-1 of 1
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Previous page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Next page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
